location-service/src/main/java/com/movie/recommender/location/service/LocationService.java

```diff
@@ -1,28 +1,83 @@
 package com.movie.recommender.location.service;
 
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
 import com.movie.recommender.location.exception.LocationNotFoundException;
+import com.movie.recommender.location.model.entity.City;
+import com.movie.recommender.location.model.entity.Country;
+import com.movie.recommender.location.repository.CityRepository;
+import com.movie.recommender.location.repository.CountryRepository;
 import com.movie.recommender.location.repository.LocationRepository;
 import com.movie.recommender.location.model.dto.LocationCreateDTO;
 import com.movie.recommender.location.model.dto.LocationUpdateDTO;
 import com.movie.recommender.location.model.entity.Location;
 import com.movie.recommender.location.model.dto.LocationDTO;
+import org.springframework.http.ResponseEntity;
 import org.springframework.stereotype.Service;
 import jakarta.transaction.Transactional;
 import lombok.extern.slf4j.Slf4j;
+import org.springframework.web.client.RestTemplate;
 
 @Slf4j
 @Service
 public class LocationService {
 
     private final LocationRepository locationRepository;
+    private final CityRepository cityRepository;
+    private final CountryRepository countryRepository;
+    private final RestTemplate restTemplate;
 
-    public LocationService(LocationRepository locationRepository) {
+    public LocationService(LocationRepository locationRepository, CityRepository cityRepository, CountryRepository countryRepository, RestTemplate restTemplate) {
         this.locationRepository = locationRepository;
+        this.cityRepository = cityRepository;
+        this.countryRepository = countryRepository;
+        this.restTemplate = restTemplate;
     }
 
     @Transactional
     public LocationDTO saveLocation(LocationCreateDTO locationCreateDTO) {
         log.info("Saving new location for user: {}", locationCreateDTO.getUserId());
+        log.info("Fetching or creating city for user: {}", locationCreateDTO.getUserId());
+
+        String cityName;
+        String countryName;
+
+        try {
+            String GEOLOCATION_URL = "https://geocode.xyz/%s,%s?geoit=json";
+            ResponseEntity<String> response = restTemplate.getForEntity(
```

Comment:
extract the geolocation functionality into a dedicated `GeolocationService`
this new service would handle all interactions with the geolocation API
in result, `GeolocationService` will work only with geolocation API requests, while `LocationService` manages location entities
this separation makes each class easier to understand and maintain